src/PIL/pdfParser.py

```diff
@@ -0,0 +1,725 @@
+        #self.orig_ref = kwargs.pop("orig_ref", None)
+        #self.pdf = kwargs.pop("pdf", None)
+        #self.is_changed = False
+
+    def __setitem__(self, key, value):
+        self.is_changed = True
+        UserDict.__setitem__(self, key, value)
+
+    def __bytes__(self):
+        #if self.orig_ref is not None:
+        #    if self.is_changed:
+        #        if self.pdf is not None:
+        #            del self.pdf.xref_table[self.orig_ref.object_id]
+        #    else:
+        #        return bytes(self.orig_ref)
+        out = bytearray(b"<<")
+        for key, value in self.items():
+            if value is None:
+                continue
+            value = pdf_repr(value)
+            out.extend(b"\n")
+            out.extend(bytes(PdfName(key)))
+            out.extend(b" ")
+            out.extend(value)
+            #out += b"\n%s %s" % (PdfName(key), value)
+        out.extend(b"\n>>")
+        return bytes(out)
+        #return out + b"\n>>"
+
+    __str__ = __bytes__
+
+    #def write(self, f, orig_ref=None, pdf=None):
+    #    self.orig_ref = orig_ref
+    #    self.pdf = pdf
+    #    f.write(bytes(self))
+
+
+class PdfBinary:
+    def __init__(self, data):
+        self.data = data
+
+    if sys.version_info.major >= 3:
+        def __bytes__(self):
+            return make_bytes("<%s>" % "".join("%02X" % b for b in self.data))
+
+        def __str__(self):
+            return bytes(self).decode("us-ascii")
+
+    else:
+        def __str__(self):
+            return "<%s>" % "".join("%02X" % ord(b) for b in self.data)
+
+
+def pdf_repr(x):
+    if x is True:
+        return b"true"
+    elif x is False:
+        return b"false"
+    elif x is None:
+        return b"null"
+    elif isinstance(x, PdfName) or isinstance(x, PdfDict) or isinstance(x, PdfArray) or isinstance(x, PdfBinary):
+        return bytes(x)
+    elif isinstance(x, int):
+        return str(x).encode("us-ascii")
+    elif isinstance(x, dict):
+        return bytes(PdfDict(x))
+    elif isinstance(x, list):
+        return bytes(PdfArray(x))
+    elif isinstance(x, str) and sys.version_info.major >= 3:
+        return pdf_repr(x.encode("utf-8"))
+    elif isinstance(x, bytes):
+        return b"(" + x.replace(b"\\", b"\\\\").replace(b"(", b"\\(").replace(b")", b"\\)") + b")"  # XXX escape more chars? handle binary garbage
+    else:
+        return bytes(x)
+
+
+class PdfParser:
+    """Based on http://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/PDF32000_2008.pdf
+    Supports PDF up to 1.4
+    """
+
+    def __init__(self, filename=None, f=None, buf=None, start_offset=0):
+        self.filename = filename
+        self.buf = buf
+        self.start_offset = start_offset
+        if buf is not None:
+            self.read_pdf_info()
+        elif f is not None:
+            self.read_pdf_info_from_file(f)
+        elif filename is not None:
+            with open(filename, "rb") as f:
+                self.read_pdf_info_from_file(f)
```

Comment:
I'm a little uncomfortable here with `self.buf` as local state and not something explicitly passed into read_pdf_info, especially as there is at least one other value that may be set here, but not in the else case below.

Reply:
I'm not sure I understand. Do you mean to eliminate self.buf completely and just pass buf around in all the calls to the methods that read and parse the PDF?

Reply:
Yes. In general, I prefer to see pure functions. If that's not feasable, a set of object attributes that are consistent over the life of the object. In this case, there's a function that (silently) requires `self.buf`, and `self.buf` is either null or not, depending on side effects in other functions.

Reply:
I have now changed the interface so that the parser keeps the file open and keeps a reference to it, making an open-read-append use case much easier. I have also eliminated all the passing of the file and buffer objects.